contracts/0.8.25/vaults/VaultHub.sol

```diff
@@ -390,98 +309,195 @@ contract VaultHub is PausableUntilWithRoles {
     /// @param _amountOfShares amount of stETH shares to mint
-    /// @dev msg.sender should be vault's owner
     function mintShares(address _vault, address _recipient, uint256 _amountOfShares) external whenResumed {
-        if (_vault == address(0)) revert ZeroArgument("_vault");
-        if (_recipient == address(0)) revert ZeroArgument("_recipient");
-        if (_amountOfShares == 0) revert ZeroArgument("_amountOfShares");
+        if (_recipient == address(0)) revert ZeroArgument();
+        if (_amountOfShares == 0) revert ZeroArgument();
 
-        _vaultAuth(_vault, "mint");
+        VaultConnection storage connection = _checkConnection(_vault);
+        if (msg.sender != connection.owner) revert NotAuthorized();
 
-        VaultSocket storage socket = _connectedSocket(_vault);
+        VaultRecord storage record = _storage().records[_vault];
 
-        uint256 vaultSharesAfterMint = socket.liabilityShares + _amountOfShares;
-        uint256 shareLimit = socket.shareLimit;
-        if (vaultSharesAfterMint > shareLimit) revert ShareLimitExceeded(_vault, shareLimit);
+        uint256 vaultSharesAfterMint = record.liabilityShares + _amountOfShares;
+        if (vaultSharesAfterMint > connection.shareLimit) revert ShareLimitExceeded(_vault, connection.shareLimit);
 
-        IStakingVault vault_ = IStakingVault(_vault);
-        if (!vault_.isReportFresh()) revert VaultReportStaled(_vault);
+        if (!_isReportFresh(record)) revert VaultReportStaled(_vault);
 
-        uint256 maxMintableRatioBP = TOTAL_BASIS_POINTS - socket.reserveRatioBP;
-        uint256 maxMintableEther = (vault_.totalValue() * maxMintableRatioBP) / TOTAL_BASIS_POINTS;
+        uint256 maxMintableRatioBP = TOTAL_BASIS_POINTS - connection.reserveRatioBP;
+        uint256 maxMintableEther = (_totalValue(record) * maxMintableRatioBP) / TOTAL_BASIS_POINTS;
         uint256 stETHAfterMint = LIDO.getPooledEthBySharesRoundUp(vaultSharesAfterMint);
         if (stETHAfterMint > maxMintableEther) {
-            revert InsufficientTotalValueToMint(_vault, vault_.totalValue());
+            revert InsufficientTotalValueToMint(_vault, _totalValue(record));
         }
 
         // Calculate the minimum ETH that needs to be locked in the vault to maintain the reserve ratio
-        uint256 minLocked = (stETHAfterMint * TOTAL_BASIS_POINTS) / maxMintableRatioBP;
+        uint256 etherToLock = (stETHAfterMint * TOTAL_BASIS_POINTS) / maxMintableRatioBP;
 
-        if (minLocked > vault_.locked()) {
-            revert VaultInsufficientLocked(_vault, vault_.locked(), minLocked);
+        if (etherToLock > record.locked) {
+            record.locked = uint128(etherToLock);
         }
 
-        socket.liabilityShares = uint96(vaultSharesAfterMint);
+        record.liabilityShares = uint96(vaultSharesAfterMint);
         LIDO.mintExternalShares(_recipient, _amountOfShares);
         OperatorGrid(LIDO_LOCATOR.operatorGrid()).onMintedShares(_vault, _amountOfShares);
```

Comment:
should it be placed upper (to fail earlier)?

be wary of the state changes against reentrancy :warning: